### DISCUSSION TITLE
XState / FSM as a backend for REST API - how?

### DISCUSSION THREAD

**Discussion by VanTanev** [statelyai/xstate#2556] [Q&A]
I'm facing a problem I don't even know how to begin to tackle:

I have a complex workflow that takes place in the backend. This workflow can be modeled as an FSM.

The workflow tracks the work being done on a particular entity, from workflow creation to worker selection, worker work, qa, publish.

This means, externally this is a REST API, but I want it to kind of behave like an FSM - for example, only specific actions are valid depending on the current state of the workflow - eg if it's in "tagging", then you can do one of "COMPLETE_TAGGING", "ABANDON_TAGGING" and "ADD_COMMENT".

Has anyone done an FSM-backed API like that? What about making it rest-compatible? Any guidance?

**Comment by davidkpiano** [comment]
Relevant Discord discussion: https://discord.com/channels/795785288994652170/799416943324823592/880090177098383360

**Comment by ChrisShank** [comment]
I think @hew has been thinking about something similar https://gist.github.com/hew/b3ed0736583556da9662f2f88bcea487

**Reply by hew** [reply]
It's not really for backend use (insofar as I can see at least). 

I actually updated it to how I'm using it at work: https://gist.github.com/hew/b3ed0736583556da9662f2f88bcea487
